libyul/AsmParser.cpp

```diff
@@ -119,99 +119,108 @@ langutil::Token Parser::advance()
-				return;
-			}
-
-			solAssert(srcTagArgsMatch.size() == 5, "");
-			position += srcTagArgsMatch.position() + srcTagArgsMatch.length();
-
-			if (srcTagArgsMatch[4].matched && (
-				!boost::algorithm::ends_with(srcTagArgsMatch[4].str(), "\"") ||
-				boost::algorithm::ends_with(srcTagArgsMatch[4].str(), "\\\"")
-			))
-			{
-				m_errorReporter.syntaxError(
-					1544_error,
-					commentLocation,
-					"Invalid code snippet in source location mapping. Quote is not terminated."
-				);
-				return;
-			}
-
-			optional<int> const sourceIndex = toInt(srcTagArgsMatch[1].str());
-			optional<int> const start = toInt(srcTagArgsMatch[2].str());
-			optional<int> const end = toInt(srcTagArgsMatch[3].str());
-
-			m_debugDataOverride = DebugData::create();
-			if (!sourceIndex.has_value() || !start.has_value() || !end.has_value())
-				m_errorReporter.syntaxError(
-					6367_error,
-					commentLocation,
-					"Invalid value in source location mapping. "
-					"Expected non-negative integer values or -1 for source index and location."
-				);
-			else if (sourceIndex == -1)
-				m_debugDataOverride = DebugData::create(SourceLocation{start.value(), end.value(), nullptr});
-			else if (!(sourceIndex >= 0 && m_sourceNames->count(static_cast<unsigned>(sourceIndex.value()))))
-				m_errorReporter.syntaxError(
-					2674_error,
-					commentLocation,
-					"Invalid source mapping. Source index not defined via @use-src."
-				);
+			if (auto parseResult = parseSrcComment(commentLiteral, m_scanner->currentCommentLocation()))
+				tie(commentLiteral, sourceLocation) = *parseResult;
 			else
-			{
-				shared_ptr<string const> sourceName = m_sourceNames->at(static_cast<unsigned>(sourceIndex.value()));
-				solAssert(sourceName, "");
-				m_debugDataOverride = DebugData::create(SourceLocation{start.value(), end.value(), move(sourceName)});
-			}
+				break;
 		}
 		else
 			// Ignore unrecognized tags.
 			continue;
 	}
+
+	m_debugDataOverride = DebugData::create(sourceLocation);
+}
+
+optional<pair<string_view, SourceLocation>> Parser::parseSrcComment(
+	string_view _arguments,
+	langutil::SourceLocation const& _commentLocation
+)
+{
+	static regex const argsRegex = regex(
+		R"~~(^(-1|\d+):(-1|\d+):(-1|\d+)(?:\s+|$))~~"  // index and location, e.g.: 1:234:-1
+		R"~~(("(?:[^"\\]|\\.)*"?)?)~~",                // optional code snippet, e.g.: "string memory s = \"abc\";..."
+		regex_constants::ECMAScript | regex_constants::optimize
+	);
+	match_results<string_view::const_iterator> match;
+	if (!regex_search(_arguments.cbegin(), _arguments.cend(), match, argsRegex))
+	{
+		m_errorReporter.syntaxError(
+			8387_error,
+			_commentLocation,
+			"Invalid values in source location mapping. Could not parse location specification."
+		);
+		return nullopt;
+	}
+
+	solAssert(match.size() == 5, "");
+	_arguments = _arguments.substr(static_cast<size_t>(match.position() + match.length()));
```

Comment:
`_arguments` is really a confusing name for it when you cut out the actual arguments from it and keep the name :) It's more like `remainingComment`. Or maybe this:
```suggestion
	string_view tail = _argumentsAndTail.substr(static_cast<size_t>(match.position() + match.length()));
```

Reply:
I'll keep `_arguments` but introduce `tail`.